### PR TITLE
PP-9675 implement search disputes service

### PR DESCRIPTION
**Pull request by SandorArpa**
## WHAT YOU DID
- implement a new service that searches Ledger for disputes
- add tests

### CODE REVIEW

File: src/main/java/uk/gov/pay/api/service/LedgerService.java [hunk 1]
```diff
@@ -118,6 +120,27 @@ public SearchRefundsResponseFromLedger searchRefunds(Account account, Map<String
...
+                .target(ledgerUriGenerator.transactionsURIWithParams(paramsAsMap))
+                .request()
+                .accept(MediaType.APPLICATION_JSON_TYPE)
+                .get();
+
+        if (response.getStatus() == SC_OK) {
+            try {
+                return response.readEntity(SearchDisputeResponseFromLedger.class);
+            } catch (ProcessingException exception) {
+                throw new SearchRefundsException(exception);
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/api/service/LedgerService.java:137]: Should this be SearchDisputesException ?

File: src/main/java/uk/gov/pay/api/service/LedgerService.java [hunk 2]
```diff
@@ -118,6 +120,27 @@ public SearchRefundsResponseFromLedger searchRefunds(Account account, Map<String
...
+            }
+        }
+
+        throw new SearchRefundsException(response);
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/api/service/LedgerService.java:141]: Same here

File: src/test/java/uk/gov/pay/api/service/SearchDisputesServiceTest.java [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+        assertThat(dispute.getPaymentId(), is("parent-abcde-12345"));
+        assertThat(dispute.getCreatedDate(), is("2022-05-20T19:05:00.000Z"));
+        assertThat(dispute.getEvidenceDueDate(), is("2022-05-27T19:05:00.000Z"));
+        assertThat(dispute.getSettlementSummary().getSettledDate(), is("2022-05-27"));
+        assertThat(dispute.getStatus(), is("lost"));
+        assertThat(dispute.getLinks().getPayment().getHref(), is("http://publicapi.test.localhost/v1/payments/parent-abcde-12345"));
+    }
+
+    @Test
+    public void getSearchResponse_shouldThrowRefundsValidationExceptionWhenParamsAreInvalid() {
```
Comment (kbottla) [inline src/test/java/uk/gov/pay/api/service/SearchDisputesServiceTest.java:90]:
```suggestion
    public void getSearchResponse_shouldThrowDisputesValidationExceptionWhenParamsAreInvalid() {
```

File: src/main/java/uk/gov/pay/api/service/LedgerService.java (hunk 2 above)
Comment (SandorArpa) [reply src/main/java/uk/gov/pay/api/service/LedgerService.java:141]: 👍

File: src/main/java/uk/gov/pay/api/service/LedgerService.java (hunk 1 above)
Comment (SandorArpa) [reply src/main/java/uk/gov/pay/api/service/LedgerService.java:137]: 👍

File: src/test/java/uk/gov/pay/api/service/SearchDisputesServiceTest.java (hunk 3 above)
Comment (SandorArpa) [reply src/test/java/uk/gov/pay/api/service/SearchDisputesServiceTest.java:90]: 👍

File: src/main/java/uk/gov/pay/api/service/SearchDisputesService.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+                                PublicApiUriGenerator publicApiUriGenerator,
+                                PaginationDecorator paginationDecorator) {
+
+        this.ledgerService = ledgerService;
+        this.publicApiUriGenerator = publicApiUriGenerator;
+        this.paginationDecorator = paginationDecorator;
+    }
+
+    public DisputeSearchResults searchDisputes(Account account, DisputesParams params) {
+        validateDisputeParameters(params);
```
[the remark this answers is not in the document]
Comment (SandorArpa) [reply src/main/java/uk/gov/pay/api/service/SearchDisputesService.java:33]: `TransactionSearchService`, `PaymentSearchService`, and `SearchRefundsService` do validation at the same point 🤔 I can refactor all four in another iteration

[the remark this answers is not in the document]
Comment (kbottla) [reply src/main/java/uk/gov/pay/api/service/SearchDisputesService.java:33]: It's fine to leave the rest as is. For disputes, validation can be moved to resource or at least need to make the method name clear (validationAndSearchDisputes or similar)

[the remark this answers is not in the document]
Comment (SandorArpa) [reply src/main/java/uk/gov/pay/api/service/SearchDisputesService.java:33]: 👍

File: src/main/java/uk/gov/pay/api/exception/SearchDisputeException.java [hunk 5]
```diff
@@ -0,0 +1,14 @@
+package uk.gov.pay.api.exception;
+
+import javax.ws.rs.core.Response;
+
+public class SearchDisputeException extends ConnectorResponseErrorException {
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/api/exception/SearchDisputeException.java:5]: This needs an exception mapper ?

**Review by kbottla — APPROVED** [review]
We also need `SearchDisputeExceptionMapper` for `SearchDisputeException ` and can be in a separate PR.

**Pull request merged**